### DISCUSSION TITLE
Some posts don't generate excerpts

### DISCUSSION THREAD

**Discussion by CharliePoole** [statiqdev/Discussions#19] [Q&A] [answered]
I'm rewriting my personal website. Started with the wyam blog recipe and had pretty good results, but it seemed more complex than it had to be so I decided to give statiq a try.

My top-level index displays a list of posts nicely... __except__ for a single post that would not show an excerpt. In an effort to figure out the cause, I set up some test posts. Now I'm seeing that even simple text does not generate an excerpt. All of these things were working for me in wyam, so I must be missing something.

Hoping for some suggestions as to what the problem might be. Are there differences in how excerpts are extracted in statiq?

Beyond that, I'm thinking of setting up in a way that I can debug right into the statiq code in order to figure this out. Any pointers on how to set that up?

**Comment by CharliePoole** [comment]
Hmmm... I just realized that the posts without excerpts are all `cshtml` files. Do I need to add something to the default pipelines?

**Comment by daveaglick** [comment, marked as answer]
It's a chicken-and-egg problem. The behavior you're seeing is actually baked into the Content pipeline (by this point in the code, Markdown has been evaluated into HTML, but Razor has not):

![image](https://user-images.githubusercontent.com/1020407/94354501-a2455f00-0049-11eb-9ed8-cd30942306da.png)

The problem is that Razor isn't actually HTML, so trying to use an HTML library like AngleSharp to parse a `.cshtml` file as if it were and then find the first `<p>` is brittle at best, and was causing lots of bug reports. Since Razor (and other templating languages) can also use code, the excerpts for Razor files tended to be goofy even when they did work (they would contain things like `@Html.Whatever()` right in the excerpt content). It only kind of worked in Wyam if you got lucky and the Razor didn't happen to blow up AngleSharp.

The ordering comes into play because we could always try to find excerpts after the Razor engine evaluates a `.cshtml` file and converts it into HTML, but by then it's too late because the whole point of the excerpt is to make it available to other Razor files and we have no way of creating a reliable per-document dependency tree (it's even worse because a blog post might have a sidebar with the excerpts from other blog posts and vice-versa, making the tree cyclic if we don't do excerpts for all documents beforehand).

The best solution would be to remove that HTML filter, or at least let it also get excerpts for Razor, but I'd need a reliable way to do that for raw `.cshtml` files and I haven't found a good one. Did that make sense?

**Reply by CharliePoole** [reply]
Yes, that makes complete sense. Funny: I was blaming the content and spent a lot of time varying it without noticing that it was only the Razor files that cause a problem. As often happens, I saw it right after trying to explain the problem here. 

I guess my approach going forward will be...
1. Stick to markdown as much as possible
2. Use frontmatter to specify the excerpt as a workaround when necessary
3. Explore the statiq source to look for something better. :smile: 

Thanks for the quick response.
